File: backend/src/service/profile.service.ts
import UserArticleInteraction from "@/models/UserArticleInteraction";
import Article from "@/models/Article";
import Categories from "@/models/Categories";
import Tags from "@/models/Tags";
import { prisma } from "@/config/db";

/**
 * 获取用户收藏的文章列表
 * @param userId 用户ID
 * @param skip 跳过的记录数
 * @param size 每页大小
 * @returns 收藏的文章列表，包含图片路径、标题、分类和发布时间
 */
export async function getArticlesCollectedId(userId: number, skip: number, size: number) {
    const articlesCollectedId = await UserArticleInteraction.find({ userId, isCollected: true })
        .skip(skip)
        .limit(size)
        .select('articleId')

    const articlesCollectedIdArr = articlesCollectedId.map(item => item.articleId)

    const articles = await Article.find({ _id: { $in: articlesCollectedIdArr } })
        .skip(skip)
        .limit(size)
        .select('cover title category published tag content').lean()

    const categoryIds = [...new Set(articles.map(a => a.category).filter(Boolean))];
    const tagIds = [...new Set(articles.map(a => a.tag).filter(Boolean))];

    const activeQuery = { deleted: { $ne: true }, status: 'ACTIVE' }

    const [categorys, tags] = await Promise.all([
        Categories.find({ _id: { $in: categoryIds }, ...activeQuery })
            .select('_id name'),
        Tags.find({ _id: { $in: tagIds }, ...activeQuery })
            .select('_id name'),
    ])

    const categoryMap = new Map(categorys.map(c => [c._id.toString(), c.name]));
    const tagMap = new Map(tags.map(c => [c._id.toString(), c.name]));

    const formattedArticles = articles
        .map(article => {
            // 提取纯文本内容并截取前30字
            let contentPreview = article.content || ''

            // 移除 HTML 标签，只保留纯文本
            contentPreview = contentPreview.replace(/<[^>]+>/g, '')

            // 截取前30个字符，如果超过则添加省略号
            contentPreview = contentPreview.length > 30
                ? contentPreview.substring(0, 30)
                : contentPreview

            return {
                ...article,
                category: categoryMap.get(article.category?.toString()) || '',
                tag: tagMap.get(article.tag?.toString()) || '',
                content: contentPreview  // 替换为摘要内容
            }
        });

    return formattedArticles
}

/**
 * 获取用户个人信息
 * @param userId 用户ID
 * @returns 用户个人信息，包含用户名、签名、头像等
 */
export async function getProfile(userId: number) {
    const userInfo = await prisma.user.findUnique({
        where: {
            user_id: userId
        },
        select: {
            username: true,
            signature: true,
            cover: true,
            phone: true,
            publish_time: true,
            update_time: true,
        }
    });

    return userInfo;
}

/**
 * 更新用户个人信息
 * @param userId 用户ID
 * @param updateData 要更新的数据
 * @returns 更新后的用户个人信息
 */
export async function updateProfile(userId: number, updateData: {
    username?: string;
    signature?: string;
    cover?: string;
    phone?: string;
}) {
    await prisma.user.update({
        where: {
            user_id: userId
        },
        data: {
            ...updateData,
            update_time: new Date()
        },
        select: {
            user_id: true,
            username: true,
            signature: true,
            cover: true,
            phone: true,
            publish_time: true,
            update_time: true,
        }
    });
}

/**
 * 根据关键词搜索用户收藏的文章
 * @param userId 用户ID
 * @param skip 跳过的记录数
 * @param size 每页大小
 * @param keyword 搜索关键词
 * @returns 符合关键词的收藏文章列表，包含图片路径、标题、分类和发布时间
 */
export async function getKeywordArticles(userId: number, skip: number, size: number, keyword: string) {
    const articlesCollectedId = await UserArticleInteraction.find({ userId, isCollected: true })
        .skip(skip)
        .limit(size)
        .select('articleId')

    const articlesCollectedIdArr = articlesCollectedId.map(item => item.articleId)

    const articles = await Article.find({
        _id: { $in: articlesCollectedIdArr },
        $or: [
            { title: { $regex: keyword, $options: 'i' } },
        ]
    })
        .skip(skip)
        .limit(size)
        .select('cover title category published').lean()

    const categoryIds = [...new Set(articles.map(a => a.category).filter(Boolean))];

    const activeQuery = { deleted: { $ne: true }, status: 'ACTIVE' }

    const categorys = await Categories.find({ _id: { $in: categoryIds }, ...activeQuery })
        .select('_id name')

    const categoryMap = new Map(categorys.map(c => [c._id.toString(), c.name]));

    const formattedArticles = articles
        .map(article => ({
            ...article,
            category: categoryMap.get(article.category?.toString()) || '',
        }));

    return formattedArticles
}

File: backend/java-backend/src/main/java/com/example/demo/service/ProfileService.java
package com.example.demo.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.example.demo.common.BusinessException;
import com.example.demo.mapper.UserMapper;
import com.example.demo.model.mongo.Article;
import com.example.demo.model.mongo.Categories;
import com.example.demo.model.mongo.Tags;
import com.example.demo.model.mongo.UserArticleInteraction;
import com.example.demo.model.mysql.User;
import com.example.demo.repository.mongo.ArticleRepository;
import com.example.demo.repository.mongo.CategoriesRepository;
import com.example.demo.repository.mongo.TagsRepository;
import com.example.demo.repository.mongo.UserArticleInteractionRepository;
import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class ProfileService {

    @Autowired
    private UserMapper userMapper;

    @Autowired
    private UserArticleInteractionRepository interactionRepository;

    @Autowired
    private ArticleRepository articleRepository;

    @Autowired
    private CategoriesRepository categoriesRepository;

    @Autowired
    private TagsRepository tagsRepository;

    @Value("${upload.dir}")
    private String uploadDir;

    @Value("${upload.base-url}")
    private String baseUrl;

    /** 查询用户个人资料 */
    public Map<String, Object> getProfile(Integer userId) {
        User user = userMapper.selectById(userId);
        if (user == null) {
            throw new BusinessException(404, "用户不存在");
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("username", user.getUsername());
        result.put("signature", user.getSignature());
        result.put("cover", user.getCover());
        result.put("phone", user.getPhone());
        result.put("publishTime", user.getPublishTime());
        result.put("updateTime", user.getUpdateTime());
        return result;
    }

    /** 更新用户资料，支持同时上传头像，返回最新资料 */
    public Map<String, Object> updateProfile(Integer userId, Map<String, String> updateData, MultipartFile avatar) {
        User user = userMapper.selectById(userId);
        if (user == null) {
            throw new BusinessException(404, "用户不存在");
        }

        if (updateData != null) {
            if (updateData.containsKey("username")) {
                user.setUsername(updateData.get("username"));
            }
            if (updateData.containsKey("signature")) {
                user.setSignature(updateData.get("signature"));
            }
            if (updateData.containsKey("phone")) {
                user.setPhone(updateData.get("phone"));
            }
        }

        if (avatar != null && !avatar.isEmpty()) {
            String originalFilename = avatar.getOriginalFilename();
            String extension = "";
            if (originalFilename != null && originalFilename.contains(".")) {
                extension = originalFilename.substring(originalFilename.lastIndexOf("."));
            }
            String filename = UUID.randomUUID().toString() + extension;

            File uploadPath = new File(uploadDir);
            if (!uploadPath.exists()) {
                uploadPath.mkdirs();
            }

            try {
                avatar.transferTo(new File(uploadPath, filename));
            } catch (IOException e) {
                throw new BusinessException(500, "文件上传失败");
            }

            String coverUrl = baseUrl + "/uploads/" + filename;
            user.setCover(coverUrl);
        }

        userMapper.updateById(user);

        return getProfile(userId);
    }

    /** 分页查询用户收藏的文章列表 */
    public Map<String, Object> getArticlesCollectedId(Integer userId, int skip, int size) {
        PageRequest pageRequest = PageRequest.of(skip / size, size);
        Page<UserArticleInteraction> interactionPage = interactionRepository
                .findByUserIdAndIsCollected(userId, true, pageRequest);

        List<UserArticleInteraction> interactions = interactionPage.getContent();
        List<String> articleIds = interactions.stream()
                .map(UserArticleInteraction::getArticleId)
                .map(ObjectId::toString)
                .collect(Collectors.toList());

        List<Map<String, Object>> articleList = new ArrayList<>();
        if (!articleIds.isEmpty()) {
            List<Article> articles = articleRepository.findAllById(articleIds);

            // 批量获取分类和标签
            Set<String> categoryIds = articles.stream()
                    .map(Article::getCategory)
                    .filter(Objects::nonNull)
                    .collect(Collectors.toSet());
            Set<String> tagIds = articles.stream()
                    .map(Article::getTag)
                    .filter(Objects::nonNull)
                    .collect(Collectors.toSet());

            Map<String, String> categoryNameMap = resolveCategoryNames(categoryIds);
            Map<String, String> tagNameMap = resolveTagNames(tagIds);

            for (Article article : articles) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("id", article.getId());
                item.put("title", article.getTitle());
                item.put("cover", article.getCover());
                item.put("category", categoryNameMap.getOrDefault(article.getCategory(), ""));
                item.put("tag", tagNameMap.getOrDefault(article.getTag(), ""));
                // 去除HTML标签后截取前30字符
                String contentPreview = article.getContent() != null ? article.getContent().replaceAll("<[^>]+>", "") : "";
                if (contentPreview.length() > 30) {
                    contentPreview = contentPreview.substring(0, 30);
                }
                item.put("content", contentPreview);
                item.put("published", article.getPublished());
                articleList.add(item);
            }
        }

        Map<String, Object> result = new HashMap<>();
        result.put("list", articleList);
        result.put("total", interactionPage.getTotalElements());
        return result;
    }

    /** 在用户收藏的文章中按标题关键词搜索 */
    public Map<String, Object> getKeywordArticles(Integer userId, int skip, int size, String keyword) {
        PageRequest pageRequest = PageRequest.of(skip / size, size);
        Page<UserArticleInteraction> interactionPage = interactionRepository
                .findByUserIdAndIsCollected(userId, true, pageRequest);

        List<UserArticleInteraction> interactions = interactionPage.getContent();
        List<String> articleIds = interactions.stream()
                .map(UserArticleInteraction::getArticleId)
                .map(ObjectId::toString)
                .collect(Collectors.toList());

        List<Map<String, Object>> articleList = new ArrayList<>();
        if (!articleIds.isEmpty()) {
            List<Article> articles = articleRepository.findAllById(articleIds);

            // 按标题关键词过滤
            List<Article> filtered = articles.stream()
                    .filter(a -> a.getTitle() != null && a.getTitle().contains(keyword))
                    .collect(Collectors.toList());

            // 批量获取分类
            Set<String> categoryIds = filtered.stream()
                    .map(Article::getCategory)
                    .filter(Objects::nonNull)
                    .collect(Collectors.toSet());

            Map<String, String> categoryNameMap = resolveCategoryNames(categoryIds);

            for (Article article : filtered) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("id", article.getId());
                item.put("title", article.getTitle());
                item.put("cover", article.getCover());
                item.put("category", categoryNameMap.getOrDefault(article.getCategory(), ""));
                item.put("published", article.getPublished());
                articleList.add(item);
            }
        }

        Map<String, Object> result = new HashMap<>();
        result.put("list", articleList);
        result.put("total", articleList.size());
        return result;
    }

    /** 批量根据分类ID查询分类名映射 */
    private Map<String, String> resolveCategoryNames(Set<String> categoryIds) {
        if (categoryIds == null || categoryIds.isEmpty()) {
            return Collections.emptyMap();
        }
        List<Categories> categories = categoriesRepository.findByIdInAndDeletedNotAndStatus(
                new ArrayList<>(categoryIds), true, "ACTIVE");
        return categories.stream().collect(Collectors.toMap(Categories::getId, Categories::getName, (a, b) -> a));
    }

    /** 批量根据标签ID查询标签名映射 */
    private Map<String, String> resolveTagNames(Set<String> tagIds) {
        if (tagIds == null || tagIds.isEmpty()) {
            return Collections.emptyMap();
        }
        List<Tags> tags = tagsRepository.findByIdInAndDeletedNotAndStatus(
                new ArrayList<>(tagIds), true, "ACTIVE");
        return tags.stream().collect(Collectors.toMap(Tags::getId, Tags::getName, (a, b) -> a));
    }
}
